### PR TITLE
Deleted variables

### PR DESCRIPTION
**Pull request by ecprice**
Support deleted variables and independent try/except scopes (addressing #1045).

### CODE REVIEW

**Comment by JukkaL** [comment]
Thanks for the PR! I'll do a quick pass today and full review probably in a few days.

File: mypy/messages.py [hunk 1]
```diff
@@ -520,6 +522,14 @@ def does_not_return_value(self, void_type: Type, context: Context) -> None:
             self.fail('{} does not return a value'.format(
                 capitalize((cast(Void, void_type)).source)), context)
 
+    def deleted_as_rvalue(self, typ: DeletedType, context: Context) -> None:
+        """Report an error about using an deleted type as an rvalue."""
+        if typ.source is None:
+            s = ""
+        else:
+            s = " '{}'".format(typ.source)
+        self.fail('Using deleted variable{} as an rvalue'.format(s), context)
```
Comment (JukkaL) [inline mypy/messages.py]: The message might be a little confusing for people not familiar with PL terminology. What about rewording the message as `Can't read deleted variable 'foo'`, for example?

File: mypy/messages.py [hunk 2]
```diff
@@ -247,6 +247,8 @@ def format_simple(self, typ: Type, verbosity: int = 0) -> str:
             return 'None'
         elif isinstance(typ, AnyType):
             return '"Any"'
+        elif isinstance(typ, DeletedType):
+            return '<Deleted>'
```
Comment (JukkaL) [inline mypy/messages.py]: Minor suggestion: Maybe reword this as `<deleted>` (without capital 'D') as otherwise this may look like `Deleted` is a user-level concept, similar to `Any` for example, even though it's an internal thing only.

File: mypy/types.py [hunk 3]
```diff
@@ -699,6 +722,9 @@ def visit_none_type(self, t):
     def visit_erased_type(self, t):
         return "<Erased>"
 
+    def visit_deleted_type(self, t):
+        return "<Deleted {}>".format(t.source)
```
Comment (JukkaL) [inline mypy/types.py]: Handle `None` value specially?

**Comment by ecprice** [comment]
I've improved the error messages.  I'm not sure what your preferred workflow is; I've left it as a new commit, but would recommend that you (or I?) squash this into 0aabbdd when rebasing to master.

**Comment by JukkaL** [comment]
Thanks for updating the PR!

Keeping the new commit is slightly preferable unless it makes rebasing harder. Also, I generally try to rebase any non-trivial branches on top of master before merging to master for a more linear history, but I'm not religious about this.

I think that the changes that I pushed today are causing merge conflicts. I can look at rebasing your branch (probably at some point next week).

**Comment by JukkaL** [comment]
I merged and pushed the PR. I actually decided to recreate the PR as a single commit and amend the author due to the merge conflicts.
